ros/angel_system_nodes/angel_system_nodes/dummy_multi_task_monitor.py

```diff
@@ -0,0 +1,360 @@
+
+
+class DummyMultiTaskMonitor(Node):
+    """
+    ROS node that simulates a node that could monitor multiple tasks
+    simultaneously. The keyboard is used to advance between task steps.
+    """
+
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+        log = self.get_logger()
+
+        param_values = declare_and_get_parameters(
+            self,
+            [
+                (PARAM_CONFIG_FILE,),
+                (PARAM_TASK_STATE_TOPIC,),
+                (PARAM_QUERY_TASK_GRAPH_TOPIC,),
+            ],
+        )
+        self._config_file = param_values[PARAM_CONFIG_FILE]
+        self._task_state_topic = param_values[PARAM_TASK_STATE_TOPIC]
+        self._query_task_graph_topic = param_values[PARAM_QUERY_TASK_GRAPH_TOPIC]
+
+        # Load the task configurations
+        with open(self._config_file, "r") as f:
+            config = yaml.safe_load(f)
+
+        self._task_state_dict: Dict[int, TaskStateInformation] = {}
+        for t in config["tasks"]:
+            # Load the config for this task
+            with open(t["config_file"], "r") as f:
+                task_config = yaml.safe_load(f)
+
+            n_steps = task_config["steps"][-1]["id"]
+
+            task_state_info = TaskStateInformation(
+                f"{t['id']}_{task_config['title']}",
+                n_steps,
+            )
+
+            # Load the list of task steps
+            task_state_info.steps = [
+                step["description"] for step in task_config["steps"]
+            ]
+
+            self._task_state_dict[t["id"]] = task_state_info
+
+        log.info("Loaded the following tasks:")
+        for key, value in self._task_state_dict.items():
+            log.info(f"    Task {key}: {value.task_title}")
+
+        # Initialize ROS hooks
+        self._task_update_publisher = self.create_publisher(
+            TaskUpdate, self._task_state_topic, 1
+        )
+        self._task_graph_service = self.create_service(
+            QueryTaskGraph, self._query_task_graph_topic, self.query_task_graph_callback
+        )
+
+        # Store module for local use.
+        self._keyboard = keyboard
+        # Start the keyboard monitoring thread
+        log.info("Starting keyboard threads")
+        self._keyboard_t = Thread(target=self.monitor_keypress)
+        self._keyboard_t.daemon = True
+        self._keyboard_t.start()
+        log.info("Starting keyboard threads... done")
+
+    def _clean_skipped_cache(self, current_hmm_step_id: int) -> None:
+        """
+        Clear the error cache any IDs greater than the given.
+        """
+        s = np.asarray(list(self._steps_skipped_cache))
+        to_remove_ids = s[s > current_hmm_step_id]
+        if to_remove_ids.size:
+            self.get_logger().info(
+                f"Clearing step IDs from skipped cache: " f"{to_remove_ids.tolist()}"
+            )
+        [self._steps_skipped_cache.remove(_id) for _id in to_remove_ids]
```

Comment:
Why do this instead of a normal for-loop? It seems like this would allocate a spurious list instance that is immediately destroyed.
```suggestion
        for _id in to_remove_ids:
            self._steps_skipped_cache.remove(_id)
```